text/3621-derive-smart-pointer.md

```diff
@@ -0,0 +1,676 @@
+```
+
+## Object safety
+
+Consider the following trait:
+```rust
+trait MyTrait {
+    // Arbitrary self types is enough for this.
+    fn func(self: MySmartPointer<Self>);
+}
+
+// But this requires #[derive(SmartPointer)].
+fn call_func(value: MySmartPointer<dyn MyTrait>) {
+    value.func();
+}
+```
+You do not need `#[derive(SmartPointer)]` to declare this trait ([arbitrary
+self types][ast] is enough), but the trait will not be object safe unless you
+annotate `MySmartPointer` with `#[derive(SmartPointer)]`. If you don't, then
+the use of `dyn MyTrait` triggers the following error:
+```
+error[E0038]: the trait `MyTrait` cannot be made into an object
+  --> src/lib.rs:11:36
+   |
+8  |     fn func(self: MySmartPointer<Self>);
+   |                   -------------------- help: consider changing method `func`'s `self` parameter to be `&self`: `&Self`
+...
+11 | fn call_func(value: MySmartPointer<dyn MyTrait>) {
+   |                                    ^^^^^^^^^^^ `MyTrait` cannot be made into an object
+   |
+note: for a trait to be "object safe" it needs to allow building a vtable to allow the call to be resolvable dynamically; for more information visit <https://doc.rust-lang.org/reference/items/traits.html#object-safety>
+  --> src/lib.rs:8:19
+   |
+7  | trait MyTrait {
+   |       ------- this trait cannot be made into an object...
+8  |     fn func(self: MySmartPointer<Self>);
+   |                   ^^^^^^^^^^^^^^^^^^^^ ...because method `func`'s `self` parameter cannot be dispatched on
+```
+Note that using the `self: MySmartPointer<Self>` syntax requires that you
+implement `Receiver` (or `Deref`), as the derive macro does not emit an
+implementation of `Receiver`.
+
+## Requirements for using the macro
+
+Whenever a `self: MySmartPointer<Self>` method is called on a trait object, the
+compiler will convert from `MySmartPointer<dyn MyTrait>` to
+`MySmartPointer<MyStruct>` using something similar to a transmute. Because of
+this, there are strict requirements on the layout of `MySmartPointer`. It is
+required that `MySmartPointer` is a struct, and that (other than one-aligned,
+zero-sized fields) it must have exactly one field. The type must either be a
+standard library pointer type (reference, raw pointer, NonNull, Box, Arc, etc.)
+or another user-defined type also using this derive macro.
+```rust
+#[derive(SmartPointer)]
+struct MySmartPointer<T: ?Sized> {
+    ptr: Box<T>,
+    _phantom: PhantomData<T>,
+}
+```
+
+### Multiple type parameters
+
+If the type has multiple type parameters, then you must explicitly specify
+which one should be used for dynamic dispatch. For example:
+```rust
+#[derive(SmartPointer)]
+struct MySmartPointer<#[pointee] T: ?Sized, U> {
+    ptr: Box<T>,
+    _phantom: PhantomData<U>,
+}
+```
+Specifying `#[pointee]` when the struct has only one type parameter is allowed,
+but not required.
```

Comment:
to avoid the same issue blocking rust-lang/libs-team#334 (conflict with third-party proc macros), could it just always assume the first `?Sized` parameter to be the pointee and get rid of the `#[pointee]` attribute?

Reply:
The first type parameter also has the issue that it is incompatible with type parameter defaults, that is `struct MySmartPointer<T, Pointee: ?Sized = u32> { ... }` wouldn't work then, since the type parameter default force it to come last. 

The `?Sized` bound as a choice is brittle, as the macro can only syntactically disambiguate, yet one could rename `use Sized as VerySized;` and do `struct MySmartPointer<T: ?VerySized> { ... }`, and more importantly it's also incompatible with trait aliases (once we get those) used in the bound which could alias `?Sized`.

Reply:
I added a section about conflicts with third-party macros in https://github.com/rust-lang/rfcs/pull/3621/commits/4cd0f76661f37a07f870a4d10a1f2c3ceacc006d.